plan/eliminate_projection.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+)
+
+// EliminateProjection eliminate projection operator to avoid the cost of memory copy in the iterator of projection.
+func EliminateProjection(p LogicalPlan, reorder bool, orderedSchema expression.Schema) LogicalPlan {
+	switch plan := p.(type) {
+	case *Projection:
+		if !projectionCanBeEliminated(plan) {
+			break
+		}
+		child := p.GetChildByIndex(0).(LogicalPlan)
+		reorder = false
+		orderedSchema = make(expression.Schema, len(plan.Exprs))
+		newSchema := make(expression.Schema, len(child.GetSchema()))
+		for j, col := range child.GetSchema() {
+			for i, expr := range plan.Exprs {
+				if col.FromID == expr.(*expression.Column).FromID && col.Position == expr.(*expression.Column).Position {
```

Comment:
1. It's ugly to convert expr to column everytime.
2. correlated column should be treated as constant.